0.2.7 refactors

modmuss50 opened the pull request:
This PR contains a bit of cleanup to loom, as well as almost drop in support for access winders, and fixes for idea 2020.

This has the abbility to generate project based jars.

modmuss50 committed:
Rough work on project based jars, skeleton for AccessEscalators?

modmuss50 committed:
First working draft

modmuss50 commented:
Ok, I have something mostly working (not tested a lot, inner classes dont work for example).

I have been testing with the following file:

```
ae\v1	v1	named
public	class	net/minecraft/client/MinecraftClient
public	field	net/minecraft/client/MinecraftClient	resourcePackDir	Ljava/io/File;
protected	method	net/minecraft/client/MinecraftClient	initializeSearchableContainers	()V
public	class	net/minecraft/client/render/chunk/ChunkBuilder$BuiltChunk$SortTask
public	field	net/minecraft/client/render/chunk/ChunkBuilder$BuiltChunk$SortTask	data	Lnet/minecraft/client/render/chunk/ChunkBuilder$ChunkData;
```

Loom remaps this and places it in META-INF of the jar:

```
ae\v1	intermediary
public	class	net/minecraft/class_846$class_851$class_4579
public	class	net/minecraft/class_310
protected	method	net/minecraft/class_310	method_1546	()V
public	field	net/minecraft/class_310	field_1757	Ljava/io/File;
public	field	net/minecraft/class_846$class_851$class_4579	field_20841	Lnet/minecraft/class_846$class_849;
```

The format is still wide open for changes, im already not sure on forcing tabs as my ide sometimes changes them to tabs.

dblsaiko commented:
> im already not sure on forcing tabs as my ide sometimes changes them to tabs.

Why not accept any amount of whitespace as a seperator?

modmuss50 committed:
Minor changes

modmuss50 commented:
Ok, done that for now, Im still not super happy with the file format. Right now there are no limits on what you can make public, people seem undecided on allowing methods/fields to only be protected.

modmuss50 committed:
Add support for mutable, better error checking when parsing file.
Code cleanup
Remap if needed when reading

dblsaiko commented on src/main/java/net/fabricmc/loom/processors/AccessEscalator.java:
```diff
@@ -0,0 +1,338 @@
+			switch (split[1]) {
+			case "class":
+				if (split.length != 3) {
+					throw new RuntimeException("Failed to parse access escalator. at line:" + line);
```
Instead of aborting immediately, I'd add the encountered errors to a list and continue trying to parse the file, then abort after it's done, that way you can see all the errors at once (and on top of that you don't get a completely irrelevant stack trace except for the message)
I'd also make the error more descriptive, something like
```
syntax error: expected '<class name> <access level>' after 'class'
at /path/to/inputfile.txt:40:
    |
 40 | class sdas as das das d asd as
    |       ^~~~~~~~~~~~~~~~~~~~~~~~
```

Same goes for all the other parse errors.

modmuss50 replied:
I can easily make the erros more descriptive, im not sure if its worth the effort to go to make it super detailed.

dblsaiko replied:
Yeah it doesn't have to be this detailed, one line error messages would definitely be enough, the main thing I would want is that it doesn't immediately throw when it encounters the first error.

modmuss50 committed:
Fix inner classes and genSources

modmuss50 commented:
Updated file format:

```
ae\v1	named
# Comments like this are supported
public	class	net/minecraft/client/MinecraftClient # And comments can be put here as well
public	field	net/minecraft/client/MinecraftClient	resourcePackDir	Ljava/io/File;
protected	method	net/minecraft/client/MinecraftClient	initializeSearchableContainers	()V
mutable	field	net/minecraft/client/MinecraftClient	COMPLETED_UNIT_FUTURE	Ljava/util/concurrent/CompletableFuture;
public	field	net/minecraft/client/MinecraftClient	COMPLETED_UNIT_FUTURE	Ljava/util/concurrent/CompletableFuture;

public	class	net/minecraft/client/render/chunk/ChunkBuilder$BuiltChunk$SortTask
public	field	net/minecraft/client/render/chunk/ChunkBuilder$BuiltChunk$SortTask	data	Lnet/minecraft/client/render/chunk/ChunkBuilder$ChunkData;
```

Tabs are required, it will complain quite quickly if you use spaces.

modmuss50 committed:
Fix CME

modmuss50 committed:
Caching, only regen jar when input changes

modmuss50 committed:
Some work, untested

modmuss50 committed:
Fix writing, fix checkstyle issues

modmuss50 committed:
More fixes

modmuss50 committed:
Move jars into a maven file structure, cleans up the file structure, and will benefit idea 2020
Add some basic validation to the AccessWidenerRemapper, will present any issues with the mappings when building (May need a way to disable?)
+ Some bugs fixes

modmuss50 committed:
Fix issues with source jars in idea 2020, should be backwards compatible with 2019

modmuss50 committed:
Move to lorenz-tiny

modmuss50 commented:
Im going to move out the AccessWinder stuff into another PR, as my reactors should be good enough for 0.2.7 even without them.

modmuss50 committed:
Build fix + small cleanup

modmuss50 committed:
Remove accesswidener's for now

modmuss50 committed:
Update dev launch injector, should fix all issues with spaces in the path.

Player3324 reviewed (approved):
No obvious issues

liach commented:
Access changer should be ok as we don't recompile decompiled code like the other silly project does

liach reviewed (approved):
No major problems
